File: Data Structures & Algorithms/matrixDFS/submission-8.py
class Solution:
    def countPaths(self, grid: List[List[int]]) -> int:
        if not grid:
            return 0
            
        rows = len(grid)
        cols = len(grid[0])

        if grid[0][0] == 1 or grid[rows - 1][cols - 1] == 1:
            return 0

        seen = set()
        paths = 0
        directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]

        def dfs(r, c):
            nonlocal paths
            if min(r, c) < 0 or r >= rows or c >= cols or grid[r][c] == 1 or (r, c) in seen:
                return

            if r == rows - 1 and c == cols - 1:
                paths += 1
                return

            seen.add((r, c))
            for dr, dc in directions:
                dfs(r + dr, c + dc)
            
            seen.remove((r, c))
            
        dfs(0, 0)
        return paths

File: Data Structures & Algorithms/count-number-of-islands/submission-8.js
class Solution {
    /**
     * @param {character[][]} grid
     * @return {number}
     */
    numIslands(grid) {
        if (!grid.length) {
            return 0
        }

        const ROWS = grid.length
        const COLS = grid[0].length
        let islands = 0

        for (let r = 0; r < ROWS; r++) {
            for (let c = 0; c < COLS; c++) {
                if (grid[r][c] == '1') {
                    this.sinkIsland(grid, r, c, ROWS, COLS)
                    islands++
                }
            }
        }
        return islands
    }

    sinkIsland(grid, r, c, m, n) {
        if (Math.min(r, c) < 0 || r >= m || c >= n || grid[r][c] == '0') return

        grid[r][c] = '0'
        const directions = [[1, 0], [0, 1], [-1, 0], [0, -1]]

        for (const [dr, dc] of directions) {
            this.sinkIsland(grid, r + dr, c + dc, m, n)
        }
    }
}
